util/ranger/new_refiner.go

```diff
@@ -0,0 +1,271 @@
+			break
+		}
+		accessInAndEqCount++
+		accessConds = append(accessConds, conditions[accessIdx])
+		if lengths[curIndex] != types.UnspecifiedLength {
+			filterConds = append(filterConds, conditions[accessIdx])
+		}
+		conditions = append(conditions[:accessIdx], conditions[accessIdx+1:]...)
+	}
+	// If curIndex equals to len of index columns, it means the rest conditions haven't been appended to filter conditions.
+	if curIndex == len(cols) {
+		filterConds = append(filterConds, conditions...)
+	}
+	return accessConds, filterConds, accessEqualCount, accessInAndEqCount
+}
+
+// buildColumnRange builds the range for sampling histogram to calculate the row count.
+func buildColumnRange(conds []expression.Expression, sc *variable.StatementContext, tp *types.FieldType) ([]*types.ColumnRange, error) {
+	if len(conds) == 0 {
+		return []*types.ColumnRange{{Low: types.Datum{}, High: types.MaxValueDatum()}}, nil
+	}
+
+	rb := Builder{Sc: sc}
+	rangePoints := FullRange
+	for _, cond := range conds {
+		rangePoints = rb.intersection(rangePoints, rb.build(cond))
+		if rb.err != nil {
+			return nil, errors.Trace(rb.err)
+		}
+	}
+	ranges := rb.buildColumnRanges(rangePoints, tp)
+	if rb.err != nil {
+		return nil, errors.Trace(rb.err)
+	}
+	return ranges, nil
+}
+
+// BuildRange is a method which can calculate IntColumnRange, ColumnRange, IndexRange.
+func BuildRange(sc *variable.StatementContext, conds []expression.Expression, rangeType int, cols []*expression.Column, lengths []int) (retRanges []types.Range,
+	accessConditions, otherConditions []expression.Expression, _ error) {
+	if rangeType == IntRangeType {
+		accessConditions, otherConditions = DetachColumnConditions(conds, cols[0].ColName)
+		ranges, err := BuildTableRange(accessConditions, sc)
+		if err != nil {
+			return nil, nil, nil, errors.Trace(err)
+		}
+		retRanges = make([]types.Range, 0, len(ranges))
+		for _, ran := range ranges {
+			retRanges = append(retRanges, ran)
+		}
+	} else if rangeType == ColumnRangeType {
+		accessConditions, otherConditions = DetachColumnConditions(conds, cols[0].ColName)
+		ranges, err := buildColumnRange(accessConditions, sc, cols[0].RetType)
+		if err != nil {
+			return nil, nil, nil, errors.Trace(err)
+		}
+		retRanges = make([]types.Range, 0, len(ranges))
+		for _, ran := range ranges {
+			retRanges = append(retRanges, ran)
+		}
+	} else if rangeType == IndexRangeType {
+		var eqAndInCount int
+		accessConditions, otherConditions, _, eqAndInCount = detachIndexScanConditions(conds, cols, lengths)
+		ranges, err := buildIndexRange(sc, cols, lengths, eqAndInCount, accessConditions)
+		if err != nil {
+			return nil, nil, nil, errors.Trace(err)
+		}
+		retRanges = make([]types.Range, 0, len(ranges))
+		for _, ran := range ranges {
+			retRanges = append(retRanges, ran)
+		}
+	}
+	return
+}
+
+// RangeSlice2IntRangeSlice changes []types.Range to []types.IntColumnRange
+func RangeSlice2IntRangeSlice(ranges []types.Range) []types.IntColumnRange {
```

Comment:
This function's name is too long. I think Ranges2IntRanges is ok. The following are same.